ecs_pod_creator.go

```diff
@@ -470,6 +488,93 @@ func (s *SecretOptions) Validate() error {
+	// resource name of an existing secret.
+	SecretName *string
+	// NewCreds are the new credentials to be stored. If this is unspecified,
+	// the secrets are assumed to already exist.
+	NewCreds *StoredRepositoryCredentials
+	// Owned determines whether or not the secret is owned by its pod or not.
+	Owned *bool
+}
+
+// NewRepositoryCredentials returns a new uninitialized set of repository
+// credentials.
+func NewRepositoryCredentials() *RepositoryCredentials {
+	return &RepositoryCredentials{}
+}
+
+// SetSecretName sets the either the friendly name for the new secret to be
+// created or the resource name of an existing secret.
+func (c *RepositoryCredentials) SetSecretName(name string) *RepositoryCredentials {
+	c.SecretName = &name
+	return c
+}
+
+// SetNewCredentials sets the new credentials to be store.
+func (c *RepositoryCredentials) SetNewCredentials(creds StoredRepositoryCredentials) *RepositoryCredentials {
+	c.NewCreds = &creds
+	return c
+}
+
+// SetOwned sets whether or not the secret credentials are owned by its pod or
+// not.
+func (c *RepositoryCredentials) SetOwned(owned bool) *RepositoryCredentials {
+	c.Owned = &owned
+	return c
+}
+
+// Validate check that the secret options are given and that either the
+// new credentials to create are specified, or the secret already exists.
+func (c *RepositoryCredentials) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(c.SecretName == nil, "must specify a secret name")
+	catcher.NewWhen(c.SecretName != nil && *c.SecretName == "", "cannot specify an empty secret name")
+	if c.NewCreds != nil {
+		catcher.Wrap(c.NewCreds.Validate(), "invalid credentials to create")
+	}
+	return catcher.Resolve()
+}
+
+// StoredRepositoryCredentials represents the storage format of repository
+// credentials for using images from private repositories.
+type StoredRepositoryCredentials struct {
+	// Username is the username for authentication.
+	Username *string `json:"username"`
+	// Password is the password for authentication.
+	Password *string `json:"password"`
+}
+
+// NewStoredRepositoryCredentials returns a new uninitialized set of repository
+// credentials for storage.
+func NewStoredRepositoryCredentials() *StoredRepositoryCredentials {
+	return &StoredRepositoryCredentials{}
+}
+
+// SetUsername sets the stored repository credential's username.
+func (c *StoredRepositoryCredentials) SetUsername(name string) *StoredRepositoryCredentials {
+	c.Username = &name
+	return c
+}
+
+// SetPassword sets the stored repository credential's password.
+func (c *StoredRepositoryCredentials) SetPassword(pwd string) *StoredRepositoryCredentials {
+	c.Password = &pwd
+	return c
+}
+
+// Validate checks that both the username and password are set.
+func (c *StoredRepositoryCredentials) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(utility.FromStringPtr(c.Username) == "", "must specify a username if secret does not already exist")
```

Comment:
`StoredRepositoryCredentials` seems like its validation should stand on its own, independent of whether a secret exists

Reply:
Fair - I updated the error message.